pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -229,6 +255,266 @@ func (p *Pod) Skip() bool {
 	return false
 }
 
+func (p *Pod) groupName() string {
+	return p.Object().GetLabels()[GroupNameLabel]
+}
+
+func (p *Pod) groupTotalCount() (int, error) {
+	if p.groupName() == "" {
+		return 0, fmt.Errorf("pod doesn't have a '%s' label", GroupNameLabel)
+	}
+
+	gtcAnnotation, ok := p.Object().GetAnnotations()[GroupTotalCountAnnotation]
+	if !ok {
+		return 0, fmt.Errorf("failed to extract '%s' annotation",
+			GroupTotalCountAnnotation)
+	}
+
+	gtc, err := strconv.Atoi(gtcAnnotation)
+	if err != nil {
+		return 0, err
+	}
+
+	if gtc < 1 {
+		return 0, fmt.Errorf("incorrect annotation value '%s=%s': group total count should be greater than zero",
+			GroupTotalCountAnnotation, gtcAnnotation)
+	}
+
+	return gtc, nil
+}
+
+func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet, error) {
+	groupTotalCount, err := p.groupTotalCount()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		resultPodSets []kueue.PodSet
+	)
+
+	for _, podInGroup := range podsInGroup.Items {
+		tc, err := strconv.Atoi(podInGroup.GetAnnotations()[GroupTotalCountAnnotation])
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract '%s' annotation from the pod '%s': %w",
+				GroupTotalCountAnnotation,
+				podInGroup.GetName(),
+				err)
+		}
+		if tc != groupTotalCount {
+			// ToDo: Add a warning event
+			return nil, fmt.Errorf("pods '%s' and '%s' has different '%s' values: %d!=%d",
+				p.GetName(), podInGroup.GetName(),
+				GroupTotalCountAnnotation,
+				groupTotalCount, tc)
+		}
+
+		roleHash, ok := podInGroup.Annotations[RoleHashAnnotation]
+		if !ok {
+			roleHash, err = getRoleHash(fromObject(&podInGroup))
+		}
+
+		roleHash = roleHash[:8]
+
+		podRoleFound := false
+		for psi := range resultPodSets {
+			if resultPodSets[psi].Name == roleHash {
+				podRoleFound = true
+				resultPodSets[psi].Count++
+			}
+		}
+
+		if !podRoleFound {
+			podSet := fromObject(&podInGroup).PodSets()
+			podSet[0].Name = roleHash
+
+			resultPodSets = append(resultPodSets, podSet[0])
+		}
```

Comment:
if there are more than 8 roles, Workload creation will fail. This will be an unrecoverable error.

Reply:
Added a pod set length validation in https://github.com/kubernetes-sigs/kueue/pull/1319/commits/fb5a4eb591521e5e3e2d7acfb4b6268a697b52e9